Investigation: Stale build cache can cause compilation errors in CI/development

## Problem

During investigation of issue #1876 (aggregate NULL handling bugs), encountered a compilation error that was blocking all testing:

```
error[E0432]: unresolved import `super::index_optimization::try_index_based_where_filtering`
 --> crates/vibesql-executor/src/select/executor/nonagg.rs:5:26
  |
5 |     index_optimization::{try_index_based_where_filtering, try_index_for_in_clause, try_spatial_index_optimization},
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `try_index_based_where_filtering` in `select::executor::index_optimization`
```

## Root Cause

The error was caused by **stale build cache**, not actual code issues. The function `try_index_based_where_filtering` exists and is properly exported in `crates/vibesql-executor/src/select/executor/index_optimization/mod.rs:16`.

## Fix

Running `cargo clean && cargo build --release` resolved the compilation error immediately.

## Impact

- Blocked investigation of #1876 until discovered
- Similar stale cache issues may affect other developers or CI runs
- Can waste significant debugging time when the actual code is correct

## Recommendation

Consider one or more of:

1. **Documentation**: Add troubleshooting section to README/CONTRIBUTING about running `cargo clean` when encountering unexplained compilation errors
2. **CI improvement**: Ensure CI runs start with clean build cache or have periodic clean builds
3. **Makefile/script**: Provide convenience script (e.g., `./scripts/clean-build.sh`) that handles cargo clean + build

## Context

- Worktree: `.loom/worktrees/issue-1876`
- Branch: `feature/issue-1876`  
- File: `crates/vibesql-executor/src/select/executor/nonagg.rs:5`
- Related: #1876

## Testing Done

After running `cargo clean && cargo build --release`:
- ✅ Build succeeded with only warnings
- ✅ Could proceed with aggregate NULL handling investigation
- ✅ Tests could run properly

--- a/README.md
+++ b/README.md
@@ -1077,6 +1077,85 @@ vibesql -c "\save production.sql"
 
 ---
 
+## 🔧 Troubleshooting
+
+### Build Errors and Compilation Issues
+
+If you encounter unexplained compilation errors, especially after switching branches or pulling updates, the issue may be caused by **stale build cache**.
+
+#### Symptoms of Stale Build Cache
+
+- Compilation errors about missing imports or functions that clearly exist
+- Errors like `unresolved import` for functions that are properly exported
+- Build failures that don't match the actual code state
+- Inconsistent build behavior across different worktrees
+
+#### Quick Fix: Clean Build
+
+The fastest way to resolve stale cache issues:
+
+```bash
+# Clean and rebuild (debug mode)
+cargo clean && cargo build
+
+# Clean and rebuild (release mode)
+cargo clean && cargo build --release
+
+# Or use the convenience script
+./scripts/clean-build.sh --release
+```
+
+#### Using the Clean Build Script
+
+The `scripts/clean-build.sh` script provides a user-friendly way to perform clean builds:
+
+```bash
+# Basic usage - clean and build in debug mode
+./scripts/clean-build.sh
+
+# Clean and build in release mode
+./scripts/clean-build.sh --release
+
+# Clean, build, and run tests
+./scripts/clean-build.sh --release --test
+
+# Verbose output for debugging
+./scripts/clean-build.sh --release --verbose
+
+# Show help and all options
+./scripts/clean-build.sh --help
+```
+
+**Script Features**:
+- Color-coded progress output
+- Multiple build modes (debug/release)
+- Optional test execution
+- Verbose mode for detailed cargo output
+- Clear error reporting
+
+#### When to Clean Build
+
+Consider running a clean build when:
+- Switching between feature branches with significant changes
+- After pulling major updates from main
+- When encountering unexplained compilation errors
+- After updating dependencies in Cargo.toml
+- Working with git worktrees (cache conflicts between worktrees)
+
+#### CI/CD Considerations
+
+CI builds should periodically use clean builds to avoid cache-related failures:
+
+```yaml
+# GitHub Actions example
+- name: Clean build
+  run: cargo clean && cargo build --release
+```
+
+For more troubleshooting help, see the [issue tracker](https://github.com/rjwalters/vibesql/issues).
+
+---
+
 ## 📖 Documentation
 
 **User Guides**:

--- a/scripts/clean-build.sh
+++ b/scripts/clean-build.sh
@@ -1,0 +1,136 @@
+#!/bin/bash
+set -euo pipefail
+
+# Clean build script for VibeSQL
+# This script helps resolve stale build cache issues by cleaning and rebuilding the project
+# Usage: ./scripts/clean-build.sh [OPTIONS]
+
+# Color output
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+YELLOW='\033[1;33m'
+BLUE='\033[0;34m'
+NC='\033[0m' # No Color
+
+# Default options
+BUILD_RELEASE=false
+RUN_TESTS=false
+VERBOSE=false
+
+# Parse command line arguments
+while [[ $# -gt 0 ]]; do
+    case $1 in
+        -r|--release)
+            BUILD_RELEASE=true
+            shift
+            ;;
+        -t|--test)
+            RUN_TESTS=true
+            shift
+            ;;
+        -v|--verbose)
+            VERBOSE=true
+            shift
+            ;;
+        -h|--help)
+            echo "Usage: $0 [OPTIONS]"
+            echo ""
+            echo "Options:"
+            echo "  -r, --release    Build in release mode (default: debug)"
+            echo "  -t, --test       Run tests after building"
+            echo "  -v, --verbose    Show verbose cargo output"
+            echo "  -h, --help       Show this help message"
+            echo ""
+            echo "Examples:"
+            echo "  $0                      # Clean and build debug"
+            echo "  $0 --release            # Clean and build release"
+            echo "  $0 --release --test     # Clean, build release, and run tests"
+            exit 0
+            ;;
+        *)
+            echo -e "${RED}Error: Unknown option: $1${NC}"
+            echo "Use --help for usage information"
+            exit 1
+            ;;
+    esac
+done
+
+# Print header
+echo -e "${BLUE}========================================${NC}"
+echo -e "${BLUE}VibeSQL Clean Build Script${NC}"
+echo -e "${BLUE}========================================${NC}"
+echo ""
+
+# Step 1: Clean build artifacts
+echo -e "${YELLOW}[1/3] Cleaning build artifacts...${NC}"
+if [ "$VERBOSE" = true ]; then
+    cargo clean
+else
+    cargo clean > /dev/null 2>&1
+fi
+echo -e "${GREEN}✓ Build artifacts cleaned${NC}"
+echo ""
+
+# Step 2: Build project
+if [ "$BUILD_RELEASE" = true ]; then
+    echo -e "${YELLOW}[2/3] Building project (release mode)...${NC}"
+    if [ "$VERBOSE" = true ]; then
+        cargo build --release
+    else
+        echo "This may take several minutes..."
+        cargo build --release 2>&1 | grep -E "(Compiling|Finished|error|warning:)" || true
+    fi
+    BUILD_PATH="target/release"
+else
+    echo -e "${YELLOW}[2/3] Building project (debug mode)...${NC}"
+    if [ "$VERBOSE" = true ]; then
+        cargo build
+    else
+        echo "This may take several minutes..."
+        cargo build 2>&1 | grep -E "(Compiling|Finished|error|warning:)" || true
+    fi
+    BUILD_PATH="target/debug"
+fi
+
+# Check if build succeeded
+if [ $? -eq 0 ]; then
+    echo -e "${GREEN}✓ Build completed successfully${NC}"
+else
+    echo -e "${RED}✗ Build failed${NC}"
+    exit 1
+fi
+echo ""
+
+# Step 3: Run tests (optional)
+if [ "$RUN_TESTS" = true ]; then
+    echo -e "${YELLOW}[3/3] Running tests...${NC}"
+    if [ "$BUILD_RELEASE" = true ]; then
+        cargo test --release
+    else
+        cargo test
+    fi
+
+    if [ $? -eq 0 ]; then
+        echo -e "${GREEN}✓ All tests passed${NC}"
+    else
+        echo -e "${RED}✗ Some tests failed${NC}"
+        exit 1
+    fi
+else
+    echo -e "${BLUE}[3/3] Skipping tests (use --test to run tests)${NC}"
+fi
+echo ""
+
+# Print summary
+echo -e "${BLUE}========================================${NC}"
+echo -e "${GREEN}Clean build completed successfully!${NC}"
+echo -e "${BLUE}========================================${NC}"
+echo ""
+echo "Build artifacts location: $BUILD_PATH"
+echo ""
+echo "Next steps:"
+if [ "$RUN_TESTS" = false ]; then
+    echo "  - Run tests: cargo test"
+fi
+echo "  - Run vibesql: $BUILD_PATH/vibesql"
+echo ""

